libpromises/loading.c

```diff
@@ -253,6 +255,44 @@ static void ShowContext(EvalContext *ctx)
     SeqDestroy(soft_contexts);
 }
 
+static void RenameMainBundle(EvalContext *ctx, GenericAgentConfig *config, Policy *policy)
+{
+    const char *const entry_point = EvalContextGetEntryPoint(ctx);
+    Seq *bundles = policy->bundles;
+    int length = SeqLength(bundles);
+    int main_counter = 0;
+    bool removed = false;
+    for (int i = 0; i < length; ++i)
+    {
+        Bundle *const bundle = SeqAt(bundles, i);
+        const char *name = bundle->name;
+        if (StringSafeEqual(name, "__main__"))
+        {
+            const char *const path = bundle->source_path;
+            char *abspath = GetAbsolutePath(path);
+            if (StringSafeEqual(abspath, entry_point))
+            {
+                Log(LOG_LEVEL_VERBOSE, "Redefining __main__ bundle from file %s to be main", abspath);
+                free(bundle->name);
+                name = bundle->name = xstrdup("main");
+                main_counter += 1;
+            }
+            else
+            {
+                Log(LOG_LEVEL_VERBOSE, "Dropping __main__ bundle from file %s (entry point: %s)", abspath, entry_point);
+                removed = true;
+                SeqSet(bundles, i, NULL);
```

Comment:
Shouldn't the bundle be properly destroyed (deallocated) here?

Reply:
SeqSet calls destroy function, so no, that would be a double free.